src/stan-users-guide/time-series.qmd

```diff
@@ -634,334 +634,116 @@ time of the original iterations.
-  // ...
-  int<lower=1> T_unsup;             // num unsupervised items
-  array[T_unsup] int<lower=1, upper=V> u;  // unsup words
-  // ...
+  // Increment target by log p(y | mu, sigma, Gamma, rho)
+  target += hmm_marginal(log_omega, gamma, rho);
 }
 ```
 
-The model for the supervised data does not change; the unsupervised
-data are handled with the following Stan implementation of the forward
-algorithm.
+The last function `hmm_marginal` takes in all the ingredients of the HMM,
+and computes the relevant log marginal distribution, $\log p(y \mid \phi)$.
 
-```stan
-model {
-  // ...
-  array[K] real acc;
-  array[T_unsup, K] real gamma;
-  for (k in 1:K) {
-    gamma[1, k] = log(phi[k, u[1]]);
-  }
-  for (t in 2:T_unsup) {
-    for (k in 1:K) {
-      for (j in 1:K) {
-        acc[j] = gamma[t - 1, j] + log(theta[j, k])
-                 + log(phi[k, u[t]]);
-      }
-      gamma[t, k] = log_sum_exp(acc);
-    }
-  }
-  target += log_sum_exp(gamma[T_unsup]);
-}
-```
-
-The forward values `gamma[t, k]` are defined to be the log
-marginal probability of the inputs `u[1],...,u[t]` up to time
-`t` and the latent state being equal to `k` at time
-`t`; the previous latent states are marginalized out.  The first
-row of `gamma` is initialized by setting `gamma[1, k]` equal
-to the log probability of latent state `k` generating the first
-output `u[1]`; as before, the probability of the first latent
-state is not itself modeled.  For each subsequent time `t` and
-output `j`, the value `acc[j]` is set to the probability of
-the latent state at time `t-1` being `j`, plus the log
-transition probability from state `j` at time `t-1` to state
-`k` at time `t`, plus the log probability of the output
-`u[t]` being generated by state `k`.  The
-`log_sum_exp` operation just multiplies the probabilities for
-each prior state `j` on the log scale in an arithmetically stable
-way.
-
-The brackets provide the scope for the local variables `acc` and
-`gamma`; these could have been declared earlier, but it is
-clearer to keep their declaration near their use.
-
-
-### Predictive inference {-}
-
-Given the transition and emission parameters, $\theta_{k, k'}$ and
-$\phi_{k,v}$ and an observation sequence $u_1, \dotsc, u_T \in \{
-1, \dotsc, V \}$, the Viterbi (dynamic programming) algorithm
-computes the state sequence which is most likely to have generated the
-observed output $u$.
-
-The Viterbi algorithm can be coded in Stan in the generated quantities
-block as follows.  The predictions here is the most likely state
-sequence `y_star[1], ..., y_star[T_unsup]` underlying the
-array of observations `u[1], ..., u[T_unsup]`.  Because this
-sequence is determined from the transition probabilities
-`theta` and emission probabilities `phi`, it may be
-different from sample to sample in the posterior.
+To obtain samples from the posterior distribution of $z$,
+we use the generated quantities block and draw, for each sample $\phi$,
+a sample from $p(z \mid y, \phi)$.
```

Comment:
You don't typically need draws from $z$.  If you do need statistics involving $z$, it's often easier to compute them in expectation (i.e., take advantage of the Rao-Blackwellization you get from marginalizing).  For example, if there are low probability but non-zero $z$, they're very hard to evaluate with sampling.  There are examples in the latent discrete parameters chapter.

You can also use the forward-backward algorithm to work out marginally $p(z_n \mid y)$ and you can use Viterbi with backtracking to find the most probable z given y.

Reply:
I added "If we desire samples from the posterior distribution of $z$..."